lib/shared/addon/components/new-edit-project/component.js

```diff
@@ -0,0 +1,114 @@
+import { alias } from '@ember/object/computed';
+import { inject as service } from '@ember/service';
+import { reject, resolve, all as PromiseAll } from 'rsvp';
+import Component from '@ember/component';
+import NewOrEdit from 'ui/mixins/new-or-edit';
+import C from 'ui/utils/constants';
+import layout from './template';
+
+export default Component.extend(NewOrEdit, {
+  layout,
+  intl: service(),
+  router: service(),
+  authzStore: service('authz-store'),
+  model: null,
+
+  primaryResource: alias('model.project'),
+  memberArray: alias('model.projectRoleTemplateBindings'),
+
+  actions: {
+    cancel() {
+      this.goBack();
+    },
+    addMember(kind) {
+      this.get('memberArray').pushObject({
+        subjectKind: kind,
+        subjectName: '',
+        projectRoleTemplateId: '',
+        projectId: '',
+      });
+    },
+    removeMember(obj) {
+      this.get('memberArray').removeObject(obj);
+    },
+  },
+
+  goBack: function () {
+    this.get('router').transitionTo('/projects');
+  },
+
+  doesNameExist() {
+    const project = this.get('primaryResource');
+    const currentProjects = this.get('model.projects');
+
+    if (currentProjects.findBy('name', project.get('name'))) {
+      return true;
+    }
+
+    return false;
+  },
+
+  doseMemberNameInvalid() {
+    const members = this.get('memberArray');
+    return members.any(r => r.subjectName.length === 0);
+  },
+
+  doseMemberRoleInvalid() {
+    const members = this.get('memberArray');
+    return members.any(r => r.projectRoleTemplateId.length === 0);
+  },
+
+  validate: function () {
+    var errors = this.get('errors', errors) || [];
+
+    if ((this.get('model.project.name') || '').trim().length === 0) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.nameReq'));
+    }
+
+    if (!this.get('editing') && this.doesNameExist()) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.nameInExists'));
+    }
+
+    if (this.doseMemberNameInvalid()) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.memberNameReq'));
+    }
+
+    if (this.doseMemberRoleInvalid()) {
+      errors.push(this.get('intl').findTranslationByKey('projectsPage.new.errors.memberRoleReq'));
+    }
+
+    if (errors.length) {
+      this.set('errors', errors.uniq());
+      return false;
+    }
+    else {
+      this.set('errors', null);
+    }
+
+    return true;
+  },
+
+  doSave() {
+    return this._super.apply(this, arguments).then((project) => {
+      const projectId = project.id;
+      const members = this.get('memberArray');
+      const promises = [];
+      members.forEach(member => {
+        member.projectId = projectId;
+        const promise = this.get('authzStore').rawRequest({
```

Comment:
what's the correct way to save a list of role bindings? is there any ember store api I can use?

Reply:
You're using it right here? Make the members binding records with authzStore.createRecord({type: "projectRoleTemplateBinding"})`in addMember and `save()` then here.

We may add an action like "setMembers" because I don't like having to do N+1 saves... But this will work for now.